src/kernel/boot.c

```diff
@@ -577,39 +577,50 @@ BOOT_CODE void init_core_state(tcb_t *scheduler_action)
 }
 
 BOOT_CODE static bool_t provide_untyped_cap(
-    cap_t      root_cnode_cap,
-    bool_t     device_memory,
-    pptr_t     pptr,
-    word_t     size_bits,
-    seL4_SlotPos first_untyped_slot
+    cap_t           root_cnode_cap,
+    bool_t          is_device_memory,
+    pptr_t          pptr,
+    word_t          size_bits,
+    seL4_SlotRegion *slot_region
 )
 {
-    bool_t ret;
-    cap_t ut_cap;
-    word_t i = ndks_boot.slot_pos_cur - first_untyped_slot;
-    if (i < CONFIG_MAX_NUM_BOOTINFO_UNTYPED_CAPS) {
-        ndks_boot.bi_frame->untypedList[i] = (seL4_UntypedDesc) {
-            .paddr    = pptr_to_paddr((void *)pptr),
-            .sizeBits = size_bits,
-            .isDevice = device_memory,
-            .padding  = {0}
-        };
-        ut_cap = cap_untyped_cap_new(MAX_FREE_INDEX(size_bits),
-                                     device_memory, size_bits, pptr);
-        ret = provide_cap(root_cnode_cap, ut_cap);
-    } else {
-        printf("Kernel init: Too many untyped regions for boot info\n");
-        ret = true;
-    }
-    return ret;
+    /* Sanity check, that the whole untyped creation for a specific slot region
+     * is not split up. Since there is a start and end filed only, no gaps are
+     * possible, this would require using multiple slot region then.
+     */
+    assert(slot_region);
+    assert(slot_region->start <= ndks_boot.slot_pos_cur);
+    assert(slot_region->end = ndks_boot.slot_pos_cur);
+    word_t i = ndks_boot.slot_pos_cur - slot_region->start;
+    if (i >= ARRAY_SIZE(ndks_boot.bi_frame->untypedList)) {
+        /* The array is full. */
+        return -2;
+    }
+
+    ndks_boot.bi_frame->untypedList[i] = (seL4_UntypedDesc) {
+        .paddr    = pptr_to_paddr((void *)pptr),
+        .sizeBits = size_bits,
+        .isDevice = is_device_memory,
+        .padding  = {0}
+    };
+
+    /* This increments ndks_boot.slot_pos_cur if the cap can be provided. */
+    if (!provide_cap(root_cnode_cap,
+                     cap_untyped_cap_new(MAX_FREE_INDEX(size_bits),
+                                         is_device_memory, size_bits, pptr))) {
+        return -1;
+    }
+
+    /* Update boot info. */
+    ndks_boot.bi_frame->untyped.end = ndks_boot.slot_pos_cur;
```

Comment:
I don't think adding a slot_region parameter to this function makes sense if internally the function requires that it can only be the ndks_boot.bi_frame->untyped region.  Also because this function is writing into the `ndks_boot.bi_frame->untypedList` array which is described by one specific slot_region.

Reply:
Yes, this should be `slot_region->end = ndks_boot.slot_pos_cur`.
The update of `ndks_boot.bi_frame->untypedList`remains a hard-coded thing - unless this also becomes a parameter. However, since I made `create_untypeds_for_region()` take a `slot_region` parameter, it should pass on this to `provide_untyped_cap()` and not fall back to something hard-coed again. Furthermore, the long term thing I'm exploring is passing the `slot_region` to `provide_cap()` also (see https://github.com/axel-h/seL4/pull/47)